helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/constraints/ConstraintBasedAlgorithm.java

```diff
@@ -70,15 +71,16 @@ public OptimalAssignment calculate(ClusterModel clusterModel) throws HelixRebala
     Set<String> busyInstances =
         getBusyInstances(clusterModel.getContext().getBestPossibleAssignment().values());
 
-    // Compute overall utilization of the cluster. Capacity dimension -> total remaining capacity
-    Map<String, Integer> overallClusterRemainingCapacityMap =
-        computeFinalClusterRemainingCapacity(clusterModel);
-
     // Create a wrapper for each AssignableReplica.
-    List<AssignableReplicaWithScore> toBeAssignedReplicas =
-        clusterModel.getAssignableReplicaMap().values().stream().flatMap(Collection::stream).map(
-            replica -> new AssignableReplicaWithScore(replica, clusterModel,
-                overallClusterRemainingCapacityMap)).sorted().collect(Collectors.toList());
+    List<AssignableReplicaWithScore> toBeAssignedReplicas = new ArrayList<>();
+    for (Map.Entry<String, Set<AssignableReplica>> replicasByResource : clusterModel
+        .getAssignableReplicaMap().entrySet()) {
+      for (AssignableReplica replica : replicasByResource.getValue()) {
+        toBeAssignedReplicas.add(new AssignableReplicaWithScore(replica, clusterModel,
+            clusterModel.getContext().getEstimateUtilizationMap()));
```

Comment:
nit: personally I prefer this in a stream fashion :D, if we can remove the Exception from the constructor.